xbmc/platform/darwin/ios-common/AnnounceReceiver.mm

```diff
@@ -0,0 +1,187 @@
+    [item setValue:[player valueForKey:@"speed"] forKey:@"speed"];
+    std::string thumb = g_application.CurrentFileItem().GetArt("thumb");
+    if (!thumb.empty())
+    {
+      bool needsRecaching;
+      std::string cachedThumb(CTextureCache::GetInstance().CheckCachedImage(thumb, false, needsRecaching));
+      //LOG("thumb: %s, %s", thumb.c_str(), cachedThumb.c_str());
+      if (!cachedThumb.empty())
+      {
+        std::string thumbRealPath = CSpecialProtocol::TranslatePath(cachedThumb);
+        [item setValue:[NSString stringWithUTF8String:thumbRealPath.c_str()] forKey:@"thumb"];
+      }
+    }
+    double duration = g_application.GetTotalTime();
+    if (duration > 0)
+      [item setValue:[NSNumber numberWithDouble:duration] forKey:@"duration"];
+    [item setValue:[NSNumber numberWithDouble:g_application.GetTime()] forKey:@"elapsed"];
+    int current = g_playlistPlayer.GetCurrentSong();
+    if (current >= 0)
+    {
+      [item setValue:[NSNumber numberWithInt:current] forKey:@"current"];
+      [item setValue:[NSNumber numberWithInt:g_playlistPlayer.GetPlaylist(g_playlistPlayer.GetCurrentPlaylist()).size()] forKey:@"total"];
+    }
+    if (g_application.CurrentFileItem().HasMusicInfoTag())
+    {
+      const std::vector<std::string> &genre = g_application.CurrentFileItem().GetMusicInfoTag()->GetGenre();
+      if (!genre.empty())
+      {
+        NSMutableArray *genreArray = [[NSMutableArray alloc] initWithCapacity:genre.size()];
+        for(std::vector<std::string>::const_iterator it = genre.begin(); it != genre.end(); ++it)
+        {
+          [genreArray addObject:[NSString stringWithUTF8String:it->c_str()]];
+        }
+        [item setValue:genreArray forKey:@"genre"];
+      }
+    }
+    //LOG(@"item: %@", item.description);
+    [g_xbmcController performSelectorOnMainThread:@selector(onPlay:) withObject:item  waitUntilDone:NO];
+  }
+  else if (msg == "OnSpeedChanged" || msg == "OnPause")
+  {
+    NSDictionary *item = [dict valueForKey:@"item"];
+    NSDictionary *player = [dict valueForKey:@"player"];
+    [item setValue:[player valueForKey:@"speed"] forKey:@"speed"];
+    [item setValue:[NSNumber numberWithDouble:g_application.GetTime()] forKey:@"elapsed"];
+    //LOG(@"item: %@", item.description);
+    [g_xbmcController performSelectorOnMainThread:@selector(OnSpeedChanged:) withObject:item  waitUntilDone:NO];
+    if (msg == "OnPause")
+      [g_xbmcController performSelectorOnMainThread:@selector(onPause:) withObject:[dict valueForKey:@"item"]  waitUntilDone:NO];
+  }
+  else if (msg == "OnStop")
+  {
+    [g_xbmcController performSelectorOnMainThread:@selector(onStop:) withObject:[dict valueForKey:@"item"]  waitUntilDone:NO];
+  }
+}
+
+long sg_singleton_lock_variable = 0;
+CAnnounceReceiver* CAnnounceReceiver::m_instance = nullptr;
+CAnnounceReceiver& CAnnounceReceiver::GetInstance()
+{
+  CAtomicSpinLock lock(sg_singleton_lock_variable);
+  if (!m_instance)
+    m_instance = new CAnnounceReceiver();
```

Comment:
The pointer could be an internal static but this code was just a move of existing code to it's own file so it could be shared among iOS and tvOS builds.